### DISCUSSION TITLE
Welcome to SourceryPro Discussions!

### DISCUSSION THREAD

**Discussion by krzysztofzablocki** [krzysztofzablocki/SourceryPro-Feedback#1] [General]
## 👋 Welcome!
  Let's use Discussions as a place to connect with other members of our community and users of Sourcery Pro. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas and workflows.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

**Comment by bcapps** [comment]
Hello @krzysztofzablocki ! I'm getting an inescapable crash on launch in Sourcery Pro that happened while I was writing a custom template. Where does Sourcery Pro store custom templates on my machine? I believe once I remove the offending template it'll launch fine. This crash continues even cross re-installs.

**Reply by krzysztofzablocki** [reply]
@bcapps Oh that's interesting, I've added a safety net that shouldn't auto-select offending template if app crashes a while ago, I wonder if it broke in some update, can you send me the data? here's where to find it

all data is inside 2 folders:
`~/Library/Containers` has 2 folders for Sourcery Pro (one for extension and the other one for standalone)
`~/Library/Group Containers`

If you could compress those 3 folders and send them over to me at krzysztof.zablocki@pixle.pl I'll figure out what's causing the issue

**Reply by krzysztofzablocki** [reply]
@bcapps I submitted 1.3.2 that restores that crash prevention (I made a mistake when I refactored code for loading that disabled it) so it will de-select the offending template and will let you export it or open it without generation active so you can fix it (probably made a dead loop), should be approved soon

**Reply by krzysztofzablocki** [reply]
@bcapps it's live so you should be good to go

**Reply by bcapps** [reply]
Thanks very much, @krzysztofzablocki! Unfortunately there may be some proprietary stuff in the failing stencil that I can't send since it's for a client. I'll see if I can reduce it to a reproducible case though and send that over if I can

**Reply by krzysztofzablocki** [reply]
@bcapps no worries, you should be able to fix it in 1.3.2 now, if it's anything else than a deadlock caused by infinite recursion just let me know so I can add it to my roadmap, I'm going to limit recursion limit in StencilSwiftKit to solve this for people

**Comment by tylerjohnson10** [comment]
I'm working on a Stencil template using Sourcery Pro version 1.3.3 and I can't seem to use it without it constantly beach-balling on me.

I'm using a 2016 MBP with 16GB RAM on Mac OS 11.3.1. 

I've tried a few things to get it working including:
1. Deleting all lines that seemed to have incorrect warnings ("type.name" is invalid, for example)
2. Disabling code generation
3. Unselecting the checkboxes next to the sample templates

The only other thing that I haven't tried yet is deleting the sample templates, but I wanted to keep those around for reference.

Below is my sample code and my current stencil code.

Sample Code:
```swift
protocol MyProtocol {
    public func fetchPeople(at location: Location) -> SignalProducer<[People], Swift.Error>
    
    public func updatePeople(_ people: People) -> SignalProducer<People, Swift.Error>
}

extension MyClass: MyProtocol {
    func fetchPeople(at location: Location) -> SignalProducer<[People], Swift.Error> {
        return .init([])
    }
    
    func updatePeople(_ people: People) -> SignalProducer<People, Swift.Error> {
        return people
    }
}
```

Stencil Code:
```
{% for type in types.protocols %}
extension {{ type.name }} { 
    {% for method in type.allMethods %}
    {{ method.accessLevel }} {{ method.name }} -> {{ method.returnTypeName|replace:"SignalProducer","AnyProducer" }}
    {% endfor %}
}
{% endfor %}
```

Video:

https://user-images.githubusercontent.com/16484169/118709583-c88b5b00-b7e2-11eb-8925-ba7c8c5de5dd.mov

**Reply by tylerjohnson10** [reply]
I'd mentioned deleting some lines of code...

Here's a screenshot of what I'm referring to. It's not clear what the problem is. I'm actually getting generated code that looks as I expect it to look. I deleted the lines with warnings in case they were causing some lag in the UI. But, as you can see from the video above, removing these additional lines doesn't seem to make a difference.

<img width="1792" alt="Screen Shot 2021-05-18 at 2 19 25 PM" src="https://user-images.githubusercontent.com/16484169/118710670-3a17d900-b7e4-11eb-948b-ac7be99784c3.png">

**Reply by krzysztofzablocki** [reply]
Thanks for the report, I'll investigate after work, my first thought is the use of filter since the syntax of them required some wacky code

**Reply by krzysztofzablocki** [reply]
@tylerjohnson10 this is really weird, not seeing that on my side, could you right click the template in the list and export it to file and attach that? I wonder if maybe there is something I'm missing

**Comment by messi** [comment]
I could not find any public road map and was wondering if there are any plans to support `swifttemplate` files?
Some other ideas that I had while using Stencil Pro for auto generating `tearDown` in unit tests:
* would love to have an *Insert Into* option for *Current Cursor*. Therefore I could jump to the class below `setUp`, press my hot key and I am done.
* that said (but probably harder to implement), having a *Run against* target for the current cursor would be nice as well so that I can set my cursor anywhere within a class and the type of the class will be my scope.

**Reply by krzysztofzablocki** [reply]
swifttemplates won't work because they require swift compiler to be executed and that's not possible in context of Xcode extensions.

Can you elaborate more on what you meant by those 2 scenarios? just to make sure I'm clear on it, do you mean this:

Insert into current cursor -> that would insert the generated code where you cursor currently is (I'd assume I should only do that if your cursor is on empty line / char otherwise it would mess things up)
Run against current type -> you mean that you'd just like the template to evaluate against the type in which body your cursor is currently in? that would be a pretty straightforward one

**Reply by messi** [reply]
(sorry for the late response)
yes, you summarized both use cases exactly how i meant them :)

**Reply by sroche27r** [reply]
@krzysztofzablocki are there any plans to support swifttemplate files in the standalone app?

**Reply by krzysztofzablocki** [reply]
In cli? It supported it for a while

**Reply by sroche27r** [reply]
We are using swifttemplates in our build phases with no problem.  
I meant in the Sourcery Pro standalone app.  
I was hoping I could author scripts in there and I would prefer to write them in swift instead of stencil.  
For scripts that aren't ending up in our build phases I wouldn't mind cut/paste from the Sourcery Pro app to our code files.

You said "_supported it for a while_"  does this mean you are no longer supporting swifttemplate files in the command line?
Although I would rather not, would you recommend that I switch to using stencil files?

**Reply by krzysztofzablocki** [reply]
ah I misread, `swifttemplates` for `sourcerytemplates`, swift templates won't be supported in SourceryPro since that can't run as part of Xcode extension due to sandboxing, Everything we have for Stencil is all hand made by me in the app so the amount of effort is quite huge and if you can't use this in extension mode then it's not worth it for me

**Reply by revolter** [reply]
Couldn't Swift templates NOT be added to the Xcode extension? I want to start using Sourcery in my project, but I want to write Swift templates, not Stencil templates, and I want to be able to use SourceryPro for instant feedback, and I'm using the CLI, so I don't need the extension at all.

**Comment by mackoj** [comment]
Hi, how can I run a Sourcery Pro template on a folder ? in a CI ?

Thanks,

**Reply by krzysztofzablocki** [reply]
Sourcery CLI supports sourcerytemplates in newest version so you’d just run that

**Comment by ghost** [comment]
Hello
I switched to a new Mac and made the mistake to not unregister the old one, so am kinda stuck and can’t activate my licence on the new one.
Can you help me out please?
Thanks,

**Reply by krzysztofzablocki** [reply]
Resolved

**Comment by mikhailmulyar** [comment]
Hi! Thanks for the great tool. There are some questions and feature requests i have in my mind.

1. Is swift templates support planned?
2. Will be nice to have AppCode plugin also.
3. It will be cool to be able to create folders for templates for better management.

**Reply by krzysztofzablocki** [reply]
hey, 

1. Swift Templates wouldn't work due to Xcode extension limitation

2. maybe in the future, I need to talk to Jetbrains

3. That's doable on App side but I don't think Xcode extensions can have sub selections althought I'll check, if the API doesn't offer that how would you imagine it working in Xcode?

**Reply by mikhailmulyar** [reply]
1. It would be cool to have properly working editor for it even if extension will not work. It still can be used with cli.
2. Ok thanks
3. Same here, it can be single flat list in Xcode extension, but it still will be cool to have this option inside app only.

**Comment by exevil** [comment]
@krzysztofzablocki: Hi Krzysztof! Great app, thanks. :)

I just wonder, is there any proper way to track changes in Sourcery Pro template files across git repos since it is base64 by default? On the other hand, regular `.stencil` templates don't seem to work with Sourcery Pro directly. Could you please explain this a little bit?

**Reply by krzysztofzablocki** [reply]
the rich sourcerytemplate files need configuration and sample code, there is also a matter of file versioning since users can be on different Sourcery Pro versions vs file version, that's why the files are binary

**Reply by exevil** [reply]
> the rich sourcerytemplate files need configuration and sample code, there is also a matter of file versioning since users can be on different Sourcery Pro versions vs file version, that's why the files are binary

Understood, thanks. But isn't there a way to have a middle ground on this? Like, for example, enable limited support of .stencil and other "raw" templates with no (or embedded) configuration and no sample code, or maybe use a bundle format that contains text configuration, sample code, and the template itself inside. I assume that encoding/decoding performance is not the issue on this amount of data.

**Reply by krzysztofzablocki** [reply]
just having stencil doesn't really define how you want the generation in context of extension work? unless you mean support for it so you can edit CLI templates?

**Reply by exevil** [reply]
Yes, it'll let us use the app as a fully-funcitioning .stencil editor for both CLI templates and its own templates. 

The flexibility of using the universal format for both use cases looks tempting but pushing binaries to the repo is quite a downside, especially during the polishing phase when it's relatively easy to override each other changes. It also disables the `Synchronize` functionality in our case for the exact same reason.

**Reply by krzysztofzablocki** [reply]
how does it disable synchronisation? do you have steps to reproduce?

**Reply by exevil** [reply]
No, it's not a bug, I meant in terms of our flow. Anyways, please consider adding more git-friendly templates format for Sourcery Pro as a feature request. :)

Thanks a lot for the answers.

**Comment by Saelyria** [comment]
Hello! I'm having an issue with 1.3.4 where the editor often undoes entered text and/or the cursor jumps to EOF. I haven't been able to narrow it down to anything super specific (seems to happen arbitrarily) other than it seems to happen more often when typing curly braces/deleting text (attached a video where it removed a chunk while typing, then jumped to EOF while deleting).

Thanks - love the software otherwise!

https://user-images.githubusercontent.com/6509382/128408217-6c692eb3-f339-4b10-a89c-fd02f63d293d.mov

**Reply by krzysztofzablocki** [reply]
would you mind sharing the template in which you are seeing this? you can send it to Krzysztof.Zablocki@pixle.pl if you can't share it here

might be completer service crashing in some scenario

**Reply by Saelyria** [reply]
No problem - I actually have it on GitHub here: https://github.com/Saelyria/Artemis/blob/develop/Tests/ArtemisTests/RenderTests/RenderTest.stencil

Thanks!

**Comment by majortom64** [comment]
Do you have any plans or would you consider writing a set of templates for Vapor 4/Fluent.

There is a huge amount of boiler plate material for both of those. Nodes did a version for Vapor 2, but a current version would be great.

**Reply by krzysztofzablocki** [reply]
I don't really use those tools but if anyone wants to write templates for it I can add it to bundled templates in SourceryPro (with credit of course)

**Comment by utahwithak** [comment]
Just wanted to add another plug for supporting `.swifttemplates` in the stand alone app. I see a lot of benefits to editing, debugging, verifying existing swift template files.  Using EJS for swift templating seems odd.

Not every sourcery workflow or user of pro is using it for the extension support. 

Also, this limitation isn't mentioned anywhere in the store or webpage and was frustrating to find missing in the pro app.

**Reply by krzysztofzablocki** [reply]
I'll add a note about Stencil / EJS, that's a good shoutout, I'm not going to support swifttemplates anytime soon though if at all. Stencil is the language I recommend for most templating in Sourcery.

The complexity behind adding Stencil support to enable the kind of workflow I did with error handling etc was crazy and I had to write my own parser and all language support from scratch, swift templates are orders of magnitude more complex because it's real swift, and you no longer would be able to use extension with them because they need to compile through standard swift compiler, outside sandbox etc

**Comment by Heroex** [comment]
Hi, i bought Sourcery Pro from AppStore and i have a question please. 

Imagine that i have two protocols (in separated files):

protocol A {
   func test()
   func log()
}

protocol B: A {

}

and class:

class Logger: B {

   init(_ value: Int) {} // value is enum for few cases (value1, value2, ...)
}


Question - can I use Sourcery for method log() from protocol A only please? But in Logger.swift.

I have to create unit tests from protocol A, but with existing functions in class Logger.

Something like this: 

func test_value1_log()
func test_value2_log() 

etc.


Is it possible in this way?

Thank you for response, M

**Reply by krzysztofzablocki** [reply]
Xcode extensions can only know the code in a single file, the api is very limited, but if you integrate normal Sourcery CLI it can run templates that know whole project AST

**Reply by Heroex** [reply]
Thank you fro response. I tried to list all methods from main class Logger, but i see only method from protocol B. Is possible to get all methods from A too please? 

I have simple code:

{% for type in types.implementing.B %}
{% for method in type.allMethods %}
func test_{{method.name}}() {
    // {{type}}
        // {{method}}
}
{% endfor %}
{% endfor %}

but everything what i get i method from B and nothing about A is there

Thank you for any advice

**Reply by krzysztofzablocki** [reply]
I just tried your template and it's giving me methods from the protocol A correctly, are you using CLI or are you still trying to use SourceryPro?

**Reply by Heroex** [reply]
It works now, strange. Probably i had typo somewhere. Thank you.

About enums please. If class Logger has enum (var type: LogType) how can I access all cases please? If I print everything from type and method i see just the name (LogType), but if I call enum.cases it return just empty array. 
This is the var type
`parameters = [MethodParameter: argumentLabel = Optional("type"), name = type, typeName = LogType,`

Thank you for any advice

**Reply by krzysztofzablocki** [reply]
if sourcery know the type you can just do `variable.type.cases`, you can find [docs for sourcery AST here](http://merowing.info/Sourcery/Classes/Enum.html)

**Reply by Heroex** [reply]
If i print variable info:

// Variable: name = type, typeName = LogType, isComputed = true, isStatic = false, readAccess = public, writeAccess = , accessLevel = (read: SourceryRuntime.AccessLevel.public, write: SourceryRuntime.AccessLevel.none), isMutable = false, defaultValue = nil, annotations = [:], documentation = [], attributes = [:], modifiers = [<SourceryRuntime.Modifier: 0x600003cf76f0>], isFinal = false, isLazy = false, definedInTypeName = Optional(ChildLogger), actualDefinedInTypeName = Optional(ChildLogger)

and 


{% for variable in type.variables where variable.name == "type" %}
// {{ variable.cases }}
{% endfor %}

doesnt work

But thank you for your time, i'll check internet little more :)

**Reply by krzysztofzablocki** [reply]
you need to access variable.type.cases like I wrote above, the 'type' is a property of Variable type, not the name of the variable like in your case :)

**Reply by Heroex** [reply]
Thank you for patience. You said "if sourcery know the type", but it looks like he doesnt - see Variable info above. Its just variable for him. I tried to use variable.type.cases etc, but nothing work. I see property name, typeName, isComputed etc, but not type. The enum is in separated file - should I access it in other stencil and use include here? I've been looking into this for a few days now, so I'm probably blind already

**Reply by krzysztofzablocki** [reply]
known types are types defined in the files that sourcery scans, if you are not running sourcery on the definition of your variable type then it's `type` property will be nil since we don't know what the type AST is

**Comment by pkurzok** [comment]
Hi @krzysztofzablocki,

first of all thanks a lot for this great App. It saved me a ton of time already.

In my current project we started adopting Swift Concurrency and I wanted to generate a Protocol Mock that is using async / await. For that I created an updated Template I found in the Sourcery Repo. But somehow it won't gerenate the new keywords.

Should this be possible in theory? Or am I doing something wrong?

Best regards
Peter

**Reply by krzysztofzablocki** [reply]
@pkurzok probably still using Sourcery release that doesn't have it, I'll make an update this weekend

**Reply by pkurzok** [reply]
@krzysztofzablocki How about that update? :) Would be great if you get to it at some point.

**Reply by krzysztofzablocki** [reply]
I submitted appstore update today, will do paddle one soon too

**Reply by Pwyll28** [reply]
Hello, @krzysztofzablocki wondering if the update has been released or not ?

**Reply by krzysztofzablocki** [reply]
@Pwyll28 not yet, coordinating with Apple because I'm running into provisioning issues with new dylib distribution model

**Reply by Pwyll28** [reply]
Okay, good luck then and thanks for all the great job !

**Comment by adarovsky** [comment]
Hi!
Thank you for a great tool! I just bought Sourcery Pro from App Store and found it great. But, it's 1.3.4 compared to 1.8.2 for CLI. Is Sourcery Pro still maintained, or we should switch to CLI even for creating code snippets?

**Reply by krzysztofzablocki** [reply]
@adarovsky the version of Sourcery isn't 1.3.4 there, that's just a version of the app, it's using 1.8 if I recall correctly, there will be an update soon but I'm running into AppStore issues atm, absolutely maintained :)

**Reply by adarovsky** [reply]
Thank you! I was missing async support in Sourcery Pro, while it is in Sourcery CLI. Hope that it will appear soon :)

**Reply by dafurman** [reply]
Seconding the interest in async support! I was pretty bummed to discover the available syntax in the template editor isn't up to date, which is the main thing that interested me with Sourcery Pro.

**Reply by kikeenrique** [reply]
Another user interested in async support.

**Comment by gprasant** [comment]
Hi @krzysztofzablocki ,

Big fan of your work! I just installed Sourcery Pro. And while I was trying it out, I accidentally deleted the Protocol Mocks template. Is there a way to add this back ? I tried reinstalling the app, but I don't see it.

**Reply by krzysztofzablocki** [reply]
yep, File -> Restore Default Templates

**Comment by ocapmycap** [comment]
Really enjoying Sourcery Pro so far. In my app (App Store version) when creating a new template, the Name textfield has a bug. I have to type one letter at a time and wait a moment before typing the next letter, otherwise the cursor resets and overwrites the previously typed letter. 

Thanks!

**Comment by ocapmycap** [comment]
Another question. I'm having difficulty iterating over nested collection variables. 

My source code is this: 

```(swift)
struct Wallet: AutoNavigable {
    var destinations: [String] = ["add","detail"]
/* ...*/
}
```
and I want to produce something like this:
```
struct WalletView: View {
enum Destinations {
case add
case detail
}
/* ...*/
}
```

But I cannot seem to cast the destinations array or get at its members properly. Here is my template:

```
{% for view in types.structs where view.based.AutoNavigable %}

class {{ view.name }}Model: ObservableObject {
    {% for var in view.variables where var.isArray %}
    {% if var.name|hasPrefix: "dest" %}
        // how do I iterate over this array? default values produces ["add","detail"] but trying to iterate over default values produces:
       // _capacityAndFlags
        // _countAndFlags
    {% endif %}
        
    {% endfor %}
}   
```

Thanks for any help!

**Comment by trqhien** [comment]
Hi @krzysztofzablocki,

Thanks for the great tool. I just purchased Sourcery Pro today and ran into an issue with Xcode extension. I was playing aroung with the default templates and Xcode kept throwing the "No types found in the current selection" warning (please see the attached recording). I've followed the provided instruction, enabled Xcode extensions in System Preferences, but Sourcery Xcode extension doesn't seem to work.  How do I fix this?

https://github.com/krzysztofzablocki/SourceryPro-Feedback/assets/19741536/0c011344-5ede-4441-9523-c859b3d3c488

I'm using Xcode 14.2 and Sourcery 1.3.5
<img width="512" alt="image" src="https://github.com/krzysztofzablocki/SourceryPro-Feedback/assets/19741536/36272a18-3297-46dd-ab09-5df9a8d46d4a">

UPDATED: I've also followed the instruction from this [link](https://nshipster.com/xcode-source-extensions/?ref=merowing.info#using-pluginkit) but still no luck

<img width="632" alt="image" src="https://github.com/krzysztofzablocki/SourceryPro-Feedback/assets/19741536/fc82dd6b-cd7f-4d60-94af-f07d7739ea29">

**Comment by krzysztofzablocki** [comment]
@trqhien your template is set to 'selection' (right pane) but you are not selecting name of the type nor the code, put your cursor on the type name and it will work

**Reply by trqhien** [reply]
Thank you, it's working for me now. Didn't notice the cursor bit.
